btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+        batch_size : int
+            Desired size of image block to be incorporated into model at each update.
+        bin_factor : int
+            Factor to bin data by.
+
+        Returns
+        -------
+        n : int
+            Number of images to incorporate into model.
+        q : int
+            Number of components for model to maintain.
+        m : int
+            Size of image block to be incorporated into model at each update.
+        d : int
+            Number of features (dimension) in each image.
+        """
+        max_events = self.psi.max_events
+        benchmarking = self.benchmarking
+        downsample = self.downsample
+
+        # set n, q, and m
+        n = num_images
+        q = num_components
+        m = batch_size
+
+        if benchmarking:
+            min_n = max(int(4 * q), 40)
+            n = min(min_n, max_events)
+            m = q
+
+            print(f"In benchmarking mode, setting q = {q}, n = {n}, m = {m}.")
+        else:
+            n = min(n, max_events) if n != -1 else max_events
+            q = min(q, n)
+            m = min(m, n)
+
+        # set d
+        det_shape = self.psi.det.shape()
+        d = np.prod(det_shape).astype(int)
+
+        if downsample:
+            if det_shape[-1] % bin_factor or det_shape[-2] % bin_factor:
+                print("Invalid bin factor, toggled off downsampling.")
+                self.downsample = False
+            else:
+                d = int(d / bin_factor**2)
+
+        return n, q, m, d
+
+    def run(self):
+        """
+        Perform iPCA on run subject to initialization parameters.
+        """
+        m = self.batch_size
+        num_images = self.num_images
+
+        # initialize and prime model, if specified
+        if self.priming and not self.benchmarking:
+            img_batch = self.get_formatted_images(
+                self.num_components, 0, self.num_features
+            )
+            self.prime_model(img_batch)
+        else:
+            self.U = np.zeros((self.split_counts[self.rank], self.num_components))
+            self.S = np.ones(self.num_components)
+            self.mu = np.zeros((self.split_counts[self.rank], 1))
+            self.total_variance = np.zeros((self.split_counts[self.rank], 1))
+
+        # divide remaining number of images into batches
+        # will become redundant in a streaming setting, need to change
+        rem_imgs = num_images - self.num_incorporated_images
+        batch_sizes = np.array(
+            [m] * np.floor(rem_imgs / m).astype(int)
+            + ([rem_imgs % m] if rem_imgs % m else [])
+        )
+
+        # update model with remaining batches
+        for batch_size in batch_sizes:
+            self.fetch_and_update_model(batch_size)
+
+        if self.benchmarking:
+            self.save_interval_data()
+
+    def get_formatted_images(self, n, start_index, end_index):
+        """
+        Fetch n - x image segments from run, where x is the number of 'dead' images.
```

Comment:
`n` appears to be an argument, but what is `x` here?

Reply:
x is the number of dead images in the retrieved block that are ultimately filtered out. Since these images are never stored in isolation there is no reference to them as a variable, so x reflects this quantity in the docstring. Let me know if this should still be changed!

Reply:
ah yes got it, thanks for the clarification!